Hook up the WebProcess objects for Web Extensions and propogate some info over.

xeenon opened the pull request:
#### 0080cc8363bfda7701fe208f44a51c01385d4f44
<pre>
Hook up the WebProcess objects for Web Extensions and propogate some info over.
Inject the browser and chrome namespace objects into Web Extension JS contexts.
Implement the simple methods on browser.runtime and browser.extension that had FIXMEs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247964">https://bugs.webkit.org/show_bug.cgi?id=247964</a>

Reviewed by Brian Weinstein.

* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WebExtensionContextParameters.h:
* Source/WebKit/Shared/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/WebExtensionControllerParameters.h:
* Source/WebKit/Shared/WebExtensionControllerParameters.serialization.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::operator== const): Added for easy equality checks.
(WebKit::AuxiliaryProcessProxy::operator!= const): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load):
(WebKit::WebExtensionController::unload):
(WebKit::WebExtensionController::addPage):
(WebKit::WebExtensionController::removePage):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::operator!= const): Changed style to match others.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::operator== const): Added for easy equality checks.
(WebKit::WebExtensionContext::operator!= const): Ditto.
(WebKit::WebExtensionContext::baseURL const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::operator== const): Added for easy equality checks.
(WebKit::WebExtensionController::operator!= const): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
(WebKit::WebExtensionMatchPattern::operator!= const):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::get):
(WebKit::WebExtensionContextProxy::getOrCreate):
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
(WebKit::WebExtensionControllerProxy::get):
(WebKit::WebExtensionControllerProxy::getOrCreate):
(WebKit::WebExtensionControllerProxy::WebExtensionControllerProxy):
(WebKit::WebExtensionControllerProxy::load): Added.
(WebKit::WebExtensionControllerProxy::unload): Added.
(WebKit::WebExtensionControllerProxy::extensionContext const):
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Create the content script world.
(WebKit::WebExtensionContext::unload): Clear the content script world.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::extension):
(WebKit::WebExtensionAPINamespace::runtime):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h:
(WebKit::WebExtensionAPIObject::WebExtensionAPIObject):
(WebKit::WebExtensionAPIObject::extensionContext):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::WebExtensionCallbackHandler): Initialize m_callbackFunction and m_rejectFunction to nullptr.
(WebKit::toJSValue): Return JSValueMakeUndefined instead of nullptr for clarity.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateHeaderFile): Added &quot;#pragma once&quot;.
(_generateImplementationFile): Use #include instead of #import.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm: Added.
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame): Do the injection for pages.
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame): Do the inejction for service workers.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
(WebKit::WebExtensionControllerProxy::extensionContext const): Added various lookup methods.
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchGlobalObjectAvailable): Call webExtensionControllerProxy if it exists.
(WebKit::WebFrameLoaderClient::dispatchServiceWorkerGlobalObjectAvailable): Ditto.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::jsContextForWorld): Added DOMWrapperWorld version.
(WebKit::WebFrame::jsContextForServiceWorkerWorld): Ditto.
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::webExtensionControllerProxy const): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
(WebKit::WebExtensionAPIExtension::isPropertyAllowed): Block getURL in v3 extensions.
(WebKit::WebExtensionAPIExtension::getURL): Added implementation.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::getURL): Added implementation.
(WebKit::WebExtensionAPIRuntime::getManifest): Ditto.
(WebKit::WebExtensionAPIRuntime::runtimeIdentifier): Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h: Fix return type for getManifest().

Canonical link: <a href="https://commits.webkit.org/256747@main">https://commits.webkit.org/256747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/262e72a4c4a308e46ab5eeca70ae9c5973b55a47

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96723 "Failed to checkout and rebase branch from PR 6535") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5988 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29817 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106251 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6202 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/34724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89107 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/102961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102398 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83335 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/5/builds/100317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/27 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/24 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/4808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2244 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/1249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/40532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -153,6 +153,9 @@ class AuxiliaryProcessProxy : public ThreadSafeRefCounted<AuxiliaryProcessProxy,
     void ref() final { ThreadSafeRefCounted::ref(); }
     void deref() final { ThreadSafeRefCounted::deref(); }
 
+    bool operator==(const AuxiliaryProcessProxy& other) const { return (this == &other); }
```
Was this needed because of any of your changes?

xeenon replied:
Yeah, I loop through the processes and check equality. This makes that cleaner. https://github.com/WebKit/WebKit/commit/410c94eb6dea7893e5cefe18bcf2075b4e3860c2#diff-0241db2427147394ab62a7da6b0aa196763c1687531ccc160250c484fe0f97d4R133

b-weinstein commented on Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
```diff
@@ -69,6 +89,29 @@ WebExtensionControllerProxy::~WebExtensionControllerProxy()
+
+    m_extensionContexts.removeIf([&](auto& entry) {
+        return entry->identifier() == contextIdentifier;
+    });
```
These are cool!

b-weinstein commented on Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
```diff
@@ -107,11 +107,36 @@ void WebExtensionControllerProxy::unload(WebExtensionContextIdentifier contextId
+RefPtr<WebExtensionContextProxy> WebExtensionControllerProxy::extensionContext(WebFrame& frame, DOMWrapperWorld& world) const
+{
+    if (!world.isNormal()) {
+        auto prefix = "WebExtension-"_s;
```
Is this prefix string somewhere else too? Can it be shared?

xeenon replied:
There isn't a good place to share this. WebKit usually does not create globals for strings like this.

xeenon committed:
Hook up the WebProcess objects for Web Extensions and propogate some info over.
Inject the browser and chrome namespace objects into Web Extension JS contexts.
Implement the simple methods on browser.runtime and browser.extension that had FIXMEs.
https://bugs.webkit.org/show_bug.cgi?id=247964

Reviewed by Brian Weinstein.

* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WebExtensionContextParameters.h:
* Source/WebKit/Shared/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/WebExtensionControllerParameters.h:
* Source/WebKit/Shared/WebExtensionControllerParameters.serialization.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::operator== const): Added for easy equality checks.
(WebKit::AuxiliaryProcessProxy::operator!= const): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load):
(WebKit::WebExtensionController::unload):
(WebKit::WebExtensionController::addPage):
(WebKit::WebExtensionController::removePage):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::operator!= const): Changed style to match others.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::operator== const): Added for easy equality checks.
(WebKit::WebExtensionContext::operator!= const): Ditto.
(WebKit::WebExtensionContext::baseURL const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::operator== const): Added for easy equality checks.
(WebKit::WebExtensionController::operator!= const): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
(WebKit::WebExtensionMatchPattern::operator!= const):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::get):
(WebKit::WebExtensionContextProxy::getOrCreate):
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
(WebKit::WebExtensionControllerProxy::get):
(WebKit::WebExtensionControllerProxy::getOrCreate):
(WebKit::WebExtensionControllerProxy::WebExtensionControllerProxy):
(WebKit::WebExtensionControllerProxy::load): Added.
(WebKit::WebExtensionControllerProxy::unload): Added.
(WebKit::WebExtensionControllerProxy::extensionContext const):
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Create the content script world.
(WebKit::WebExtensionContext::unload): Clear the content script world.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::extension):
(WebKit::WebExtensionAPINamespace::runtime):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h:
(WebKit::WebExtensionAPIObject::WebExtensionAPIObject):
(WebKit::WebExtensionAPIObject::extensionContext):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::WebExtensionCallbackHandler): Initialize m_callbackFunction and m_rejectFunction to nullptr.
(WebKit::toJSValue): Return JSValueMakeUndefined instead of nullptr for clarity.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateHeaderFile): Added "#pragma once".
(_generateImplementationFile): Use #include instead of #import.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm: Added.
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame): Do the injection for pages.
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame): Do the inejction for service workers.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
(WebKit::WebExtensionControllerProxy::extensionContext const): Added various lookup methods.
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchGlobalObjectAvailable): Call webExtensionControllerProxy if it exists.
(WebKit::WebFrameLoaderClient::dispatchServiceWorkerGlobalObjectAvailable): Ditto.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::jsContextForWorld): Added DOMWrapperWorld version.
(WebKit::WebFrame::jsContextForServiceWorkerWorld): Ditto.
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::webExtensionControllerProxy const): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
(WebKit::WebExtensionAPIExtension::isPropertyAllowed): Block getURL in v3 extensions.
(WebKit::WebExtensionAPIExtension::getURL): Added implementation.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::getURL): Added implementation.
(WebKit::WebExtensionAPIRuntime::getManifest): Ditto.
(WebKit::WebExtensionAPIRuntime::runtimeIdentifier): Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h: Fix return type for getManifest().

Canonical link: https://commits.webkit.org/256747@main
